lib/uplink/bucket_attrs_test.go

```diff
@@ -0,0 +1,189 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"bytes"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/pkg/storj"
+)
+
+type testConfig struct {
+	planetCfg *testplanet.Config
+	uplinkCfg Config
+}
+
+func testPlanetWithLibUplink(t *testing.T, cfg testConfig,
+	testFunc func(*testing.T, *testcontext.Context, *testplanet.Planet, *Project)) {
+	if cfg.planetCfg == nil {
+		cfg.planetCfg = &testplanet.Config{SatelliteCount: 1, StorageNodeCount: 5, UplinkCount: 1}
+	}
+
+	planet, err := testplanet.NewCustom(zaptest.NewLogger(t), *cfg.planetCfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ctx := testcontext.New(t)
+	defer ctx.Check(planet.Shutdown)
+	defer ctx.Cleanup() // warning: this must be deferred after planet.Shutdown so it runs before
+	planet.Start(ctx)
+
+	// we only use testUplink for the free API key, until such time
+	// as testplanet makes it easy to get another way :D
+	testUplink := planet.Uplinks[0]
+	satellite := planet.Satellites[0]
+	cfg.uplinkCfg.Volatile.TLS.SkipPeerCAWhitelist = true
+
+	apiKey, err := ParseAPIKey(testUplink.APIKey[satellite.ID()])
+	if err != nil {
+		t.Fatalf("could not parse API key from testplanet: %v", err)
+	}
+	uplink, err := NewUplink(ctx, &cfg.uplinkCfg)
+	if err != nil {
```

Comment:
try using require.NoError

Reply:
stylistically, I don't like require.NoError here, because it implies it's just one of the conditions we intend to test. but this function is more of a harness: it is setting up a test environment and then running the actual test.

if it _really_ bothers you, require.NoError would work fine, but it's unhelpful here imo